src/ProductManagement.java

```diff
@@ -0,0 +1,124 @@
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class ProductManagement {
+
+  private final List<Product> catalog = new ArrayList<>();
+
+  public void addProduct(Product product) {
+    catalog.add(product);
+  }
+
+  public List<Product> getByTypeAndPrice(ProductType type, BigDecimal price) {
+    return catalog.stream()
+        .filter(p -> p.getType() == type && p.getPrice().compareTo(price) > 0)
+        .toList();
+  }
+
+  public Product getByTypeAndLowestPrice(ProductType type) {
+    return catalog.stream()
+        .filter(p -> p.getType() == type && p.getPrice() != null)
+        .min(Comparator.comparing(Product::getPrice))
+        .orElseThrow(() -> new RuntimeException("No product found"));
+  }
+
+  public List<Product> applyDiscounts() {
+    return catalog.stream()
+        .filter(product -> product.getDiscount() != null && product.getDiscount().compareTo(BigDecimal.ZERO) > 0)
+        .map(product -> product.withPrice(calcDiscount(product)))
+        .toList();
+  }
+
+  private BigDecimal calcDiscount(Product product) {
+    return product.getPrice().multiply(BigDecimal.ONE.subtract(product.getDiscount()
+        .divide(BigDecimal.valueOf(100), 2, RoundingMode.FLOOR)));
+  }
+
+  public List<Product> getLastThreeProducts() {
+    return catalog.stream()
+        .filter(p -> p.getAdded() != null)
+        .sorted(Comparator.comparing(Product::getAdded).reversed())
+        .limit(3)
+        .toList();
+  }
+
+  public BigDecimal calcOverallPriceBy(ProductType type, BigDecimal maxPrice) {
+    return catalog.stream()
+        .filter(p -> p.getAdded() != null)
+        .filter(p -> p.getType() == type && p.getPrice().compareTo(maxPrice) < 0)
+        .filter(p -> p.getAdded().isAfter(LocalDateTime.now().withDayOfYear(1).withHour(0).withMinute(0).withSecond(0)))
```

Comment:
opt: consider using more readable format:
```
  public static void main(String[] args) {
    LocalDateTime timeAddedAt = LocalDateTime.now().minusDays(1);
    LocalDateTime timeCurrentYearStartedAt = LocalDateTime.now().minusYears(1);
    System.out.println(timeAddedAt.isAfter(timeCurrentYearStartedAt));
  }
```

Reply:
sorry, can't understood, i need to change to something like this:

```
public BigDecimal calcOverallPriceBy(ProductType type, BigDecimal maxPrice) {
    LocalDateTime startOfYear = LocalDateTime.now().withDayOfYear(1).withHour(0).withMinute(0).withSecond(0);
    return catalog.stream()
        .filter(p -> p.getAdded() != null)
        .filter(p -> p.getType() == type && p.getPrice().compareTo(maxPrice) < 0)
        .filter(p -> p.getAdded().isAfter(startOfYear))
        .map(Product::getPrice)
        .reduce(BigDecimal.ZERO, BigDecimal::add);
  }
```